docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+- The two concepts could be collapsed into one if we used ephemeral node IDs that were only ever used one time,
+  but this makes it harder to efficiently handle the case of the same physical node restarting, where we may want
+  to retain the same attachments as it had before restart, without having to map them to a different node ID.
+- We may in future run the pageserver in a different cluster runtime environment (e.g. k8s) where
+  guaranteeing that the cluster manager (e.g. k8s) doesn't run two concurrent instances with the same
+  node ID is much harder.
+
+#### Distinction between attachment and node generation id
+
+The most important generation number is the tenant attachment number: this alone would be sufficient
+to implement safe tenant migration and instance failover, if we assume that our control plane will never concurrently
+run two nodes with the same node ID.
+
+Running two nodes with the same ID might sound far-fetched, but it would happen very easily if we
+ran the pageserver in a k8s environment with a StatefulSet, as k8s provides no guarantee that
+an old pod is dead before starting a new one.
+
+The node generation is useful other ways, beyond it's core correctness purpose:
+
+- Including both generations (and the node ID) in object keys (see [object keys](#object-key-changes)) also provides some flexibility in
+  how HA should work in future: we could move to a model where failover can happen within one
+  attachment generation number (i.e. without control plane coordination) without changing the
+  storage format, because node IDs/generations would de-conflict writes from peers.
+- The node generation is also useful for managing remote objects which are not per-tenant,
+  such as the persistent per-node deletion queue which is proposed in this RFC.
+- node generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+Note that the two generation numbers have a different behavior for stale generations:
+
+- A pageserver with a stale generation number should immediately terminate: it is never intended
+  to have two pageservers with the same node ID. This is not necessary for correctness, as the
+  node generation appears in object keys, but objects written in a stale node generation are never
+  read later, so waste storage.
+- An attachment with a stale generation number is permitted to continue operating (ingesting WAL
+  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
+  facilitating live migration (this will be articulated in the next RFC that describes HA and
+  migrations). Operating in a stale attachment generation is safe, although for efficiency reasons
+  should not be done over long periods of time, as the data in the stale generation has to avoid
+  doing operations involving deletion, like remote compaction.
```

Comment:
The big surprising thing in this proposal is that we need two separate generation numbers. So let's dig a little deeper into that and consider the alternatives:

1. What if there was just one generation number? Let's forget about the "node generation number", and use just "attachment generation" number.  Let's look at this section from that point of view:

> The most important generation number is the tenant attachment number: this alone would be sufficient
to implement safe tenant migration and instance failover, if we assume that our control plane will never concurrently run two nodes with the same node ID.
>
> Running two nodes with the same ID might sound far-fetched, but it would happen very easily if we
ran the pageserver in a k8s environment with a StatefulSet, as k8s provides no guarantee that
an old pod is dead before starting a new one.

Agreed, it could happen. But they would still operate independently. If the same tenant was attached to both of them, they would be assigned a different attachment number.

> The node generation is useful other ways, beyond it's core correctness purpose:
>
> - Including both generations (and the node ID) in object keys (see [object keys](#object-key-changes)) also provides some flexibility in how HA should work in future: we could move to a model where failover can happen within one attachment generation number (i.e. without control plane coordination) without changing the storage format, because node IDs/generations would de-conflict writes from peers.

An alternative is that the control plane could assign a *range* of generation numbers, and let the pageservers coordinate and divide up that range how they want.  

I guess that's another way to think about the attachment and node generation numbers. If you think of it as just one generation number, then the "attachment" part is the high bits and the "node" part is the low bits. That might be a better mental model than thinking of them as two independent numbers.

Another thought is that if we have some other system that can perform failover without the control plane, then we can move the assignment of generation numbers to that system completely. In that world, the control plane isnt't the source of truth anymore. If it wants to attach a tenant, it asks that other system to do it, and to assign a new generation number.

> - The node generation is also useful for managing remote objects which are not per-tenant,

You could assign a generation number to objects that are not tenants, too.

> such as the persistent per-node deletion queue which is proposed in this RFC.

Oh! I haven't read this proposal that far yet, but I am very surprised the deletion queues are not per-tenant. My gut reaction is that it's simpler if everything is per tenant. I'll go read that before I comment more on that..

> - node generation numbers enable building fencing into any network protocol (for example communication with safekeepers) to refuse to communicate with stale/partitioned nodes.

Didn't understand this part. You can do that with attachment generation numbers too?

Reply:
>>node generation numbers enable building fencing into any network protocol (for example communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
> Didn't understand this part. You can do that with attachment generation numbers too?

The idea was that node generations operated on an "if I'm stale, game over, I should panic" basis, whereas attachment generations didn't mean the whole node should die, just that someone else took over that tenant.  No longer relevant since I've removed node generations.

> Another thought is that if we have some other system that can perform failover without the control plane, then we can move the assignment of generation numbers to that system completely. In that world, the control plane isnt't the source of truth anymore. If it wants to attach a tenant, it asks that other system to do it, and to assign a new generation number.

Yeah, this would be the option of building a consensus system into pageservers, or giving the pageservers a new coordinator service that had its own database -- I've ruled that out here mainly because of urgency, but also because it's largely a naming thing: we can call the pageserver management code part of the control plane, or call it a separate service -- the important thing is just that there is this oracle for handing out generations, and in the short term using the control plane's database for that is expedient